File: RUBY/ncs.rb
# ncs.rb
module BaseConverter
  def self.val(c)
    if c >= '0' && c <= '9'
      c.to_i
    else
      c.upcase.ord - 'A'.ord + 10
    end
  end

  def self.to_deci(str, base)
    len = str.length
    power = 1
    num = 0
    (len - 1).downto(0) do |i|
      char_val = val(str[i])
      if char_val >= base
        puts "Invalid Number"
        return -1
      end
      num += char_val * power
      power = power * base
    end
    num
  end

  def self.re_val(num)
    if num >= 0 && num <= 9
      (num + '0'.ord).chr
    else
      (num - 10 + 'A'.ord).chr
    end
  end

  def self.from_deci(base, input_num)
    res = ""
    return "0" if input_num == 0

    while input_num > 0
      res += re_val(input_num % base)
      input_num /= base
    end
    res.reverse
  end

  def self.convert_base(s, a, b)
    num = to_deci(s, a)
    return if num == -1
    ans = from_deci(b, num)
    puts "ANSWER: #{ans}"
  end
end

File: PERL/ncs.pl
# ncs.pl

# Function to get value of a character
sub val {
    my $c = shift;
    if ($c ge '0' && $c le '9') {
        return $c; # Perl converts string '0'-'9' to int automatically
    } else {
        return (ord(uc($c)) - ord('A') + 10);
    }
}

# Function to convert a number from given base to decimal
sub toDeci {
    my ($str, $base) = @_;
    my $s = uc($str); # Convert to uppercase for consistency
    my $len = length($s);
    my $power = 1;
    my $num = 0;

    for (my $i = $len - 1; $i >= 0; $i--) {
        my $char_val = val(substr($s, $i, 1));
        if ($char_val >= $base) {
            print "Invalid Number\n";
            return -1;
        }
        $num += $char_val * $power;
        $power *= $base;
    }
    return $num;
}

# Function to get character from an integer value
sub reVal {
    my $num = shift;
    if ($num >= 0 && $num <= 9) {
        return chr($num + ord('0'));
    } else {
        return chr($num - 10 + ord('A'));
    }
}

# Function to convert a decimal number to a string in a given base
sub fromDeci {
    my ($base, $inputNum) = @_;
    my $res = "";
    if ($inputNum == 0) {
        return "0";
    }

    while ($inputNum > 0) {
        $res .= reVal($inputNum % $base);
        $inputNum = int($inputNum / $base); # Integer division
    }
    return reverse $res;
}

# Main conversion function
sub convertBase {
    my ($s, $a, $b) = @_;
    my $num = toDeci($s, $a);
    if ($num == -1) {
        return; # Handle invalid number
    }
    my $ans = fromDeci($b, $num);
    print "ANSWER: $ans\n";
}

1; # Required for 'require' to succeed
